XbSymbolDatabase.c

```diff
@@ -84,59 +89,97 @@ static inline uint32_t BitScanReverse(uint32_t value)
 
 typedef uint8_t* memptr_t;
 
-typedef bool (*custom_scan_func_t)(uint32_t LibraryFlag,
-                                   const xbe_section_header* pSectionHeader,
-                                   unsigned short BuildVersion,
-                                   const char* LibraryStr,
-                                   xb_symbol_register_t register_func,
-                                   memptr_t xb_start_virt_addr);
+typedef struct _OutputHandler {
+    xb_output_message   verbose_level;
+    xb_output_message_t func;
+} OutputHandler;
+
+// Library Type is a requirement to prevent another thread
+// doing the same scan process.
+typedef enum _eLibraryType {
+    LT_D3D = 0,
+    LT_AUDIO,
+    LT_JVS,
+    LT_XAPI,
+    LT_GRAPHIC,
+    LT_NETWORK,
+    LT_COUNT,
+    LT_UNKNOWN=LT_COUNT
+} eLibraryType;
+
+typedef enum _eScanStage {
+    SS_NONE = 0,
+    SS_1_MANUAL,
+    SS_2_SCAN_LIBS,
+} eScanStage;
+
+typedef struct _iXbSymbolContext {
+    bool                    strict_build_version_limit;
+    bool                    one_time_scan;
+    bool                    scan_first_detect;
+    uint32_t                library_filter;
+    xbaddr                  xref_database[XREF_COUNT];
+    OutputHandler           output;
+    xb_symbol_register_t    register_func;
+    XbSDBLibraryHeader      library_input;
+    XbSDBSectionHeader      section_input;
+    bool                    library_active[LT_COUNT];
+    eScanStage              scan_stage;
+#ifdef MULTI_THREAD_SAFE
+    mtx_t                   mutex;
+#endif
+} iXbSymbolContext;
+
+typedef bool (*custom_scan_func_t)(iXbSymbolContext* pContext,
+                                   const XbSDBLibrary* pLibrary,
+                                   const XbSDBSection* pSection);
 
 typedef const struct _PairScanLibSec {
     uint32_t library;
-    const char *section[PAIRSCANSEC_MAX];
+    const char* section[PAIRSCANSEC_MAX];
 } PairScanLibSec;
 
 typedef const struct _SymbolDatabaseList {
+    eLibraryType    library_type;
+    PairScanLibSec  LibSec;
 
-    PairScanLibSec LibSec;
-
-    OOVPATable     *OovpaTable;
+    OOVPATable*     OovpaTable;
     unsigned int    OovpaTableCount;
 } SymbolDatabaseList;
 
 SymbolDatabaseList SymbolDBList[] = {
     // Support inline functions in .text section
-    { XbSymbolLib_D3D8 | XbSymbolLib_D3D8LTCG,{ Sec_text, Sec_D3D, Sec_FLASHROM }, D3D8_OOVPA, D3D8_OOVPA_COUNT },
+    { LT_D3D, XbSymbolLib_D3D8 | XbSymbolLib_D3D8LTCG,{ Sec_text, Sec_D3D, Sec_FLASHROM }, D3D8_OOVPA, D3D8_OOVPA_COUNT },
```

Comment:
Why include a library type enum value here, when there's also a `internal_GetLibraryType` function that's doing the same? Either remove this variable, and use the function, or remove the function (or implement it as a lookup in SymbolDBList, instead of switch/case code)

Reply:
Okay, I'll need to take a step back to see what you're seeing. I'm sure your suggestion method is better. 😁

Reply:
After reviewing again, obviously this variable is no longer needed. 😃